examples/trl/ddpo.py

```diff
@@ -0,0 +1,248 @@
+        try:
+            cached_path = hf_hub_download(model_id, model_filename)
+        except EntryNotFoundError:
+            cached_path = os.path.join(model_id, model_filename)
+        state_dict = torch.load(cached_path, map_location=torch.device("cpu"))
+        self.mlp.load_state_dict(state_dict)
+        self.dtype = dtype
+        self.eval()
+
+    @torch.no_grad()
+    def __call__(self, images):
+        device = next(self.parameters()).device
+        inputs = self.processor(images=images, return_tensors="pt")
+        inputs = {k: v.to(self.dtype).to(device) for k, v in inputs.items()}
+        embed = self.clip.get_image_features(**inputs)
+        # normalize embedding
+        embed = embed / torch.linalg.vector_norm(embed, dim=-1, keepdim=True)
+        return self.mlp(embed).squeeze(1)
+
+
+def aesthetic_scorer(hub_model_id, model_filename, use_habana=True):
+    scorer = AestheticScorer(
+        model_id=hub_model_id,
+        model_filename=model_filename,
+        dtype=torch.float32,
+    )
+    if use_habana:
+        scorer = scorer.to("hpu")
+    else:
+        scorer = scorer.to("cpu")
+
+    def _fn(images, prompts, metadata):
+        images = (images * 255).round().clamp(0, 255).to(torch.uint8)
+        scores = scorer(images)
+        return scores, {}
+
+    return _fn
+
+
+# list of example prompts to feed stable diffusion
+animals = [
+    "cat",
+    "dog",
+    "horse",
+    "monkey",
+    "rabbit",
+    "zebra",
+    "spider",
+    "bird",
+    "sheep",
+    "deer",
+    "cow",
+    "goat",
+    "lion",
+    "frog",
+    "chicken",
+    "duck",
+    "goose",
+    "bee",
+    "pig",
+    "turkey",
+    "fly",
+    "llama",
+    "camel",
+    "bat",
+    "gorilla",
+    "hedgehog",
+    "kangaroo",
+]
+
+
+def prompt_fn():
+    return np.random.choice(animals), {}
+
+
+def image_outputs_logger(image_data, global_step, accelerate_logger):
+    # For the sake of this example, we will only log the last batch of images
+    # and associated data
+    result = {}
+    images, prompts, _, rewards, _ = image_data[-1]
+
+    for i, image in enumerate(images):
+        prompt = prompts[i]
+        reward = rewards[i].item()
+        result[f"{prompt:.25} | {reward:.2f}"] = image.unsqueeze(0).float()
+
+    accelerate_logger.log_images(
+        result,
+        step=global_step,
+    )
+
+
+if __name__ == "__main__":
+    parser = HfArgumentParser((ScriptArguments, DDPOConfig))
+    args, ddpo_config = parser.parse_args_into_dataclasses()
+    ddpo_config.mixed_precision = "bf16" if args.bf16 else "no"
+    ddpo_config.project_kwargs = {
+        "logging_dir": "./logs",
+        "automatic_checkpoint_naming": True,
+        "total_limit": 5,
+        "project_dir": "./save",
+    }
+
+    # 1. initialize Gaudi config:
+    gaudi_config = GaudiConfig.from_pretrained(args.gaudi_config_name) if args.use_habana else None
+    gaudi_config.use_torch_autocast = gaudi_config.use_torch_autocast if args.bf16 else False
```

Comment:
Should we say gaudi_config.use_torch_autocast or args.bf16?
gaudi_config.use_torch_autocast value can be changed as user setting.

Reply:
Yes, I agree. We should set
```py
gaudi_config.use_torch_autocast = gaudi_config.use_torch_autocast or args.bf16
ddpo_config.mixed_precision = "bf16" if gaudi_config.use_torch_autocast else "no"
```
That way, if the Gaudi config is saved, it will have the right value for autocast.

Reply:
@yeonsily @regisss Good catch. I had set `torch_autocast` to False due to an error when `ddpo_config.mixed_precision` was no. I fixed the issue by deleting the autocast in `modeling_sd_base.py` `__call__` which had the data type hard-coded as bfloat16. Autocast was not needed because it is set in `ddpo_trainer.py` `_generate_samples`.